pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -367,6 +398,17 @@ public void replaceSegment(ImmutableSegment segment, @Nullable ThreadSafeMutable
     Lock segmentLock = SegmentLocks.getSegmentLock(_tableNameWithType, segmentName);
     segmentLock.lock();
     try {
+      // Skip adding segments that has segment EndTime in the comparison cols earlier than (largestSeenTimestamp - TTL).
+      // Note: We only update largestSeenComparisonValue when addRecord, and access the value when addOrReplaceSegments.
+      // We only support single comparison column for TTL-enabled upsert tables.
+      if (_largestSeenComparisonValue > 0) {
```

Comment:
We cannot skip replacing segment because the new segment doesn't have snapshot, thus we don't know which docs are valid. Unfortunately I don't see a good way to handle it if user is replacing a segment that is out of TTL. Or we only allow user to upload segment without any invalid doc. We need to add more comments explaining the logic.
Also, we can move this check outside of the segment lock

Reply:
thanks @Jackie-Jiang. let me remove this check for now.
i will add a todo to allow user to skip loading segment without any invalid docs.

Reply:
This code path is invoked during segment reload also, so we need to consider that as well